### ISSUE TITLE
RF 'Automatic variables' do not seem supported when used in RED Variables mapping

### ISSUE THREAD

**Issue by GerwinLaagland** [nokia/RED#155]
I've done a search on this subject and couldn't find anything similar quite similar. Perhaps it could be related to relative pathing issues such as:
- https://github.com/nokia/RED/issues/7
- https://github.com/nokia/RED/issues/68
 
So I figured I'd do a mention, and see what comes up. The problem is related to these variables:

http://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#automatic-variables

I'm currently attempting to make use of Red, and declare a a static location variable for the use of Importing a Resource file that is used across multiple testcases. This particular location that I wish to use is a 'relative' location for which I would like to use "${EXECDIR}".

Unfortunately, the used variable from the red variable mapping isn't correctly interpreted when applied as a variable behind the resource declaration in the resource file. It is a little bit inconsistent behavior, because if I use the actual parameter _value_ with the resource declaration, there is no problem. And no validations errors are shown.

Example below:

**Variable mapping in Red**
Name: ${CREDENTIALS}
Value: ${EXECDIR}${/}Resources${/}TestSuite.RobotFramework.robot

Implemented in resource file:

```
***Setting***
Resource    ${CREDENTIALS}

```
**Returned error:**
The import name/path '${CREDENTIALS}' is parameterized. Some of used parameters cannot be resolved. Use Variable mappings in red.xml for parameter resolution

Thoughts are appreciated. :-)

**Comment by KrzysztofJozefowicz** [comment]
Hi
Variable Mappings is used to resolve parametrised path to resources and libraries only during test edit. 
I have not used Variable Mappings for providing full path to resource/lib as in your example, intent was to use variable mapping to resolve path to file and not path+filename (I need to check how robot behaves for import resources and libs resolved from string variables) .
Following your example, I would propose to change Resource import to
```
***Setting***
Resource    ${CREDENTIALS}${/}TestSuite.RobotFramework.robot
```
in Variable Mappings use:
```
Name: ${CREDENTIALS}
Value: ${EXECDIR}${/}Resources
```
Do you have exact case when you in your testcase to provide resource destination (path+resource filename) by variable? You are the first person to come up with such need.

**Comment by GerwinLaagland** [comment]
I'm unsure what is meant with provide an 'exact case'. But the goal is managability of resource files. What we're currently seeing in our project is that it's easier to import the many resources through a centralized resource file. It doesn't hamper performance on test execution to the amount that it matters.

So in order to become more flexible, we've centralized all our resources to an individual file. RED doesn't like this, and does provide warnings about nested files, but for now I believe that to be acceptable. :-)

However this file is still required to be imported in **every** testcase. Which results in the following being used in every test we have:

```
***Setting***
Resource          ${EXECDIR}${/}CustomLibrary${/}AllResources.robot
```
This particular resource then contains:

```
***Setting***
Documentation    A collection of all resources, which can be included in scripts to execute Robot functionalities.

Library    OperatingSystem
Library    DateTime
Library    Process
Library    SSHLibrary
Library    DatabaseLibrary
[etc.]

Resource    ${EXECDIR}${/}CustomLibrary${/}DomainSubject.Execute.robot
Resource    ${EXECDIR}${/}CustomLibrary${/}DomainSubject.Input.robot
Resource    ${EXECDIR}${/}CustomLibrary${/}DomainSubject.Validate.robot```
[etc.]
```
We include this file in every testcase, in order to be flexible in the setup of our tests. Now that we've started to delve into the Robot Framework a little more, we've found that the location is a little bit... well, confusing due to legacy decisions. (We're still beginners at this phase I'll admit)

Thus we wanted to parameterize the file / location in order to be more flexible for future moving about library and resource files. So that you are not required to change it everywhere in every test... which I might have to do if we do also change naming conventions on our 'AllResources.robot' file, or equivalent credential file.

Your suggestion is a good idea, which I've just tried. Unfortunately that also still results in a similar error. Which I believe to be caused by the use of ${EXECDIR}'. I've attempted an absolute location that includes a file such as:

"C:\ws\git\robot\BronReplicatieTest\Resources\Credentials.RobotFramework.robot"

And this does work.

**Comment by KrzysztofJozefowicz** [comment]
We both misunderstood each other a bit :)
Let me straight up:
- Operating-system variables and other runtime variables needs to have thiers values assigned while editing,especially for ${EXECDIR} as this unknown even for tester (value depends on where robot is called - in RED this is the path to Project)

I misread you initial question, VM can be used as you suggest although ${EXECDIR} needs to be also mapped:
```
***Setting***
Resource    ${CREDENTIALS}
```

as ${CREDENTIALS} is composed from ${EXECDIR}${/}Resources${/}TestSuite.RobotFramework.robot thus ${EXECDIR} needs to be defined in VM prior:
```
${EXECDIR} => path to workspace
${CREDENTIALS} => ${EXECDIR}${/}Resources${/}TestSuite.RobotFramework.robot
```

${EXECDIR} is not assigned by default to not produce unnecessary problems when RED would have different value than intended in suites. In our opinion it it better explicitly assigned by user than by guessing. 

I hope it helps :)

**Comment by GerwinLaagland** [comment]
Thank you for clearing this up!

Personally though, I find it odd that when I use the operating variables such "EXECDIR" inside robot script itself, the RED plugin does correctly validate the input. As it recognizes that the EXECDIR is part of the project starting location. But when used in RED VM itself, it does not. This seems to be inconsistent with the behavior of robot framework itself who supports this feature.

I assume that in your case RED is used across multiple robot instances/projects to have this feature implemented in this way? We specifically chose to use EXECDIR in our scripts because the starting point of robot execution relative to its scripts doesn't change and is a constant. The 'absolute' location where it the project is placed however,  depends on where the git clone has been placed by the developer.

To not see the RF feature repeated in the plugin is a bit unexpected. :-(

In our project we have one red instance per robot 'executable/project', and because we've also placed the red.xml file in our git repository having an absolute reference that manually has to be adjusted for developers/testers would be undesirable.

Much obliged in providing me with this information! I'll see if I can make it work with importing variable files instead for now. :-)

**Comment by KrzysztofJozefowicz** [comment]
Hi
EXECDIR can change from runtime to runtime (it is defined by place where Robot is executed) thus RED recognise EXECDIR by name but for importing libs and resources, it needs to have assigned value in VM. When RED executes tests,EXECDIR is always path to Project path but this can differ when running the same testcase in CI as it depends on robot execution command line (RED uses command: python -m robot.run --suites <dot.path.to.suites.from.project> --tests <dot.paths.to.tests> <project_path_as_starting_path> which allows to have consistent EXECDIR regardless of subfolder you run the tests from, if you use python -m robot.run <path_to_suite> EXECDIR will be different and I hope you are aware of the implications :) ). 
We will discuss the implications to assign EXECDIR path to project (per project) but we must evaluate if this would not break existing workflow (maybe this should be hidden behind preference).  Unfortunately it is not a simple case.

**Comment by GerwinLaagland** [comment]
Thank you for your elaborate response! I had not considered that you could of course start of your robot tests could be from an external source. Currently our testset is not yet part of any type of delivery pipeline so we only use the eclipse/Red run configuration to start specific folders with our test suits.

T'is sad, but no automated firing of scripts after installing packages here as of yet.  It's briljant that you'd want to have a look into an option such as preference, I did spot a few discussions about relative positioning with the red framework. One of which:

https://stackoverflow.com/questions/46748067/robot-framework-issue-when-parameterizing-resource-path-in-settings

So it does appear that I'm not the only one looking into this. Please close the issue with conclusion if it does turn out to be breaking compatibility. :-)

**Comment by KrzysztofJozefowicz** [comment]
Good news, RED already resolves ${EXECDIR} to project path per project although not in case when used as value in VM:
```
Resource  ${EXECDIR}${/}Resources${/}TestSuite.RobotFramework.robot  => will work
used in VM as ${CREDENTIALS} => ${EXECDIR}${/}Resources${/}TestSuite.RobotFramework.robot  will not work
```
We will provide enhancement as we already resolve variables mapped to values used in another mapping but ${EXECDIR} was missed. 
This should be ready in release early 2018 (most probably 0.8.3, it will not go in the release before Christmas break).

**Comment by GerwinLaagland** [comment]
Excellent news indeed! I'll be looking forward to this enhancement. Happy holidays, and thank you for your continued efforts on improving this application! :+1:

**Comment by KrzysztofJozefowicz** [comment]
released in 0.8.3
